### PR TITLE
Improve synonym generation and precompute meta infoboxes

### LINKED ISSUE #103: Precompute meta infoboxes when creating the backend
Meta infoboxes are created at runtime. The very first request about a symbol with class `wikipedia-president` takes slightly longer. Once created, the meta infobox is cached in memory. All subsequent requests about `wikipedia-president`s use the cached meta infobox.

We might as well precompute all meta infoboxes when we generate the backend. This will make WikipediaBase slightly faster.

The main benefit will actually be for other applications such as wikiscout, where we need to scan over all infoboxes and rendered attributes.

### LINKED ISSUE #109: disambiguation pages should generate synonyms, not symbols
`generate-symbols-and-synonyms` (or whatever it's called these days) doesn't realize that https://en.wikipedia.org/wiki/Sakharov is a disambiguation page, so it creates a symbol `Sakharov`, and https://en.wikipedia.org/wiki/Andrei_Sakharov doesn't get `Sakharov` as a synonym.

### LINKED ISSUE #112: bad(?) synonym "directed" -- maybe many more like it
Why do so many articles have `directed` as a synonym?  When I go to the article for `Sports director` and chase the "What links here" link, I don't find anything about "directed". Ditto "Left 4 Dead".
 
    START> (omnibase-get-symbols "directed" :class "wikibase-term")
    ((WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Sports director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Left 4 Dead" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Non-executive director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "HMS Director (1784)" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Artistic director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Music video director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Music director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Funeral director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Film director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Fibre Channel switch" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Executive director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director telephone system" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director string" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director (military)" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director-general" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director (colonial)" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director (business)" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director (band)" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Director (album)" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Cruise director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Creative director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Courage Brewery" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Chief financial officer" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Chief executive officer" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Adobe Director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Aaron Director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Tournament director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "The Director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Theatre director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Television director" :PRIORITY 0)
     (WIKIBASE-TERM "directed" :SPAN (0 8) :NAME "Technical director" :PRIORITY 0))

Bogus synonyms slow START down by making it waste time temporarily defining them as words and then undefining them (which requires calls to WikipediaBase's `get-classes`, among other work), so it's pretty important to eliminate bogus synonyms.

### LINKED ISSUE #132: Look over latest generated synonyms
I've added diff files to `/data/infolab/misc/WikipediaBase/synonyms/`. The most informative files are `symbols-deleted` and `synonyms-deleted`.

In #121, I made a change that gets rid of many bad synonyms. Here's a summary of my changes:

* Look for "May refer to:" patterns to identify disambiguation pages
* Ignore redirects that point to disambiguation pages
* Ignore symbol chains of length > 3
* Ignore redirects to local anchors
* Require disambiguation links to be only at the beginning of a sentence

Because the generation procedure takes about 40 hours, I ran it against the Dec '15 dump. Previously, synonyms were generated using the Oct '15 dump.

Totals:
* 4625630 symbols (82581 deleted, 59585 new)
* 11661793 synonyms (2892820 deleted, 381574 new)

Some seemingly correct synonyms are also deleted (I say seemingly because though semantically reasonable, they might be generated from an incorrect path along the redirect graph). However, Wikipedia is huge, and I doubt that this will have an impact on actual questions from users. Because we're biased for precision, I think this is the right move. In the future, maybe we can relax some constraints (e.g. lower the priority of synonyms with a chain longer than 3).

### PR DESCRIPTION
**Pull request by alvaromorales**
This change gets rid of many bad synonyms. Changes to the generation procedure:

* Look for "May refer to:" patterns to identify disambiguation pages
* Ignore redirects that point to disambiguation pages
* Ignore symbol chains of length > 3
* Ignore redirects to local anchors
* Require disambiguation links to be only at the beginning of a sentence

I generated symbols and synonyms on nauru with the Dec '15 dump.

4625630 symbols (82581 deleted, 59585 new)
11661793 synonyms (2892820 deleted, 381574 new)

The change gets rid of every single item with "direct" or "directing" as a synonym, yay! "Flower" is still a synonym for "Bambi" because Everything In Wikipedia Is Hard: There's a redirect "Flower (Bambi)" that redirects to "Bambi". Note that we're (correctly) not generating this synonym from the "Flower (disambiguation)" page.

I also precompute MetaInfobox objects for every valid infobox template, and save them to the backend.

Fixes #112 and fixes #103 and fixes #109 and fixes #132

### CODE REVIEW

File: scripts/create_db.py [hunk 1]
```diff
@@ -31,7 +31,7 @@ def extract_pages(f):
 
     :param file-like f: Handle on Wikipedia dump. May be any type supported by
         etree.iterparse.
-    :return (page_id, title, markup) triples
+    :return (page_id, title, markup, ns) tuple
```
Comment (michaelsilver) [inline scripts/create_db.py]: Sorry, what is this `ns` you've added all over in the `create_db.py` file? What does it stand for?

File: wikipediabase/dbutil.py [hunk 2]
```diff
@@ -32,10 +33,25 @@ class Article(Model):
     title = TextField(index=True)
     markup = TextField()
 
+    # namespace: articles (ns 0) and templates (ns 10)
+    # see https://en.wikipedia.org/wiki/Wikipedia:Namespace#Programming
+    ns = IntegerField()
```
Comment (alvaromorales) [inline wikipediabase/dbutil.py]: see above for a description of `ns`

File: README.md [hunk 3]
```diff
@@ -54,6 +54,12 @@ Install `wikipediabase`:
 
    Install [Git LFS](https://git-lfs.github.com/) and commit the changes to files in `data/`.
 
+1. Generate meta infoboxes (optional). This step takes about 4 hours.
```
Comment (michaelsilver) [inline README.md]: Maybe mention the motivating reason to generate meta infoboxes (will make WikiBase faster).

**Comment by michaelsilver** [comment]
Added a minor comment to the README, but other than that everything looks good to me! :rocket:

**Comment by alvaromorales** [comment]
Thanks for the review! I'll wait on @sfelshin to skim over the generated synonyms and for us to run a full START test suite on navassa before merging.

**Pull request merged**